ivy/functional/frontends/tensorflow/keras/metrics.py

```diff
@@ -162,6 +163,17 @@ def kl_divergence(y_true, y_pred):
 kullback_leibler_divergence = kl_divergence
 
 
+def log_cosh(y_true, y_pred):
+    y_true, y_pred = promote_types_of_tensorflow_inputs(y_true, y_pred)
+    diff = y_pred - y_true
+    logval = ivy.log(2.0)
```

Comment:
This line should be replaced by `logval = ivy.astype(ivy.log(2.0), diff.dtype)`. The problem now is that ivy's default float dtype (float64) is being used, and then typecasted to the other variables in the following line. This causes the test to fail due to type mismatch with the ground truth return.

Reply:
Yeah, that's what I've noticed when troubleshooting this problem. I set `logval = ivy.astype(ivy.log(2.0), diff.dtype)`, which causes the test case to pass if `shared_dtype=True`, but still fails if set to `False`. If I switch `y_true, y_pred = promote_types_of_tensorflow_inputs(y_true, y_pred)` to be `y_true = ivy.astype(y_true, y_pred.dtype)`, the function passes the test case with `shared_dtype` set either to `True` or `False`.

Reply:
Yes, that's right. I had a look at the original source code here https://github.com/keras-team/keras/blob/v2.10.0/keras/losses.py#L1892-L1935 and it turns out this function doesn't use tensorflow's type promotion (probably because it's the keras one, 'cause `tfp.math.log_cosh` does promote the types the way I was suggesting).
Anyway, yes, the code should be like this
`def log_cosh(y_true, y_pred):
    y_true = ivy.astype(y_true, y_pred.dtype)
    diff = y_pred - y_true
    logval = ivy.astype(ivy.log(2.0), diff.dtype)
    return ivy.mean(diff + ivy.softplus(-2.0 * diff) - logval, axis=-1)`

Reply:
Alright, switched the implementation back 👍